### PR TITLE
feat(forum): add tooltips showing absolute dates

### PR DESCRIPTION
**Pull request by wescopeland**
This PR adds `title`-driven tooltips to forum dates, offering one way for users to see an "absolute" date of a timestamp rendered in timeago format.

This affects:
* Forum posts
* Forum Activity component on the home page
* Forum Activity component on the forum index page
* Forum Post History (forumposthistory.php) page

Screenshot 2023-07-08 at 2 35 44 PM

Screenshot 2023-07-08 at 2 35 35 PM

Screenshot 2023-07-08 at 2 34 36 PM

Screenshot 2023-07-08 at 2 32 31 PM

### CODE REVIEW

File: resources/views/community/components/forum/post-comment-meta.blade.php [hunk 1]
```diff
@@ -41,5 +41,14 @@
...
 <p class='smalltext !leading-[14px]'>
     {{-- Keep this all on a single line so white space isn't added before the comma --}}
-    {{ $formattedPostTimestamp }}@if($formattedEditTimestamp)<span class='italic smalltext !leading-[14px]'>, <span class='hidden sm:inline'>last</span> edited {{ $formattedEditTimestamp }}</span> @endif
+    <span title="{{ $postCreatedTimestamp }}" class="cursor-help">{{ $formattedPostTimestamp }}@if($formattedEditTimestamp), @endif</span>
+
+    @if($formattedEditTimestamp)
+        <span class='italic smalltext !leading-[14px]'>
+            <span class='hidden sm:inline'>last</span> edited
+            <span class="cursor-help" title="{{ $postEditedTimestamp }}">
+                {{ $formattedEditTimestamp }}
```
Comment (Jamiras) [inline resources/views/community/components/forum/post-comment-meta.blade.php:50]: Doesn't really make sense to have the helper tooltip for raw dates:
image

Comment (wescopeland) [reply resources/views/community/components/forum/post-comment-meta.blade.php:70]: Took a bit of conditional logic, but should be resolved in https://github.com/RetroAchievements/RAWeb/pull/1652/commits/1c9522c294615d9fff468fecfb064c704398e134.

File: resources/views/community/components/forum/post-comment-meta.blade.php [hunk 2]
```diff
@@ -40,6 +52,23 @@
 @endif
 
 <p class='smalltext !leading-[14px]'>
-    {{-- Keep this all on a single line so white space isn't added before the comma --}}
-    {{ $formattedPostTimestamp }}@if($formattedEditTimestamp)<span class='italic smalltext !leading-[14px]'>, <span class='hidden sm:inline'>last</span> edited {{ $formattedEditTimestamp }}</span> @endif
+    <span 
+        title="{{ $shouldUsePostedTimeAgoDate ? $postCreatedTimestamp : null }}"
```
Comment (Jamiras) [inline resources/views/community/components/forum/post-comment-meta.blade.php:56]: Would it be possible to use the pretty-formatted date in to the tooltip?

image

i.e. `July 14 2023, 6:36pm`

File: resources/views/community/components/forum/post-comment-meta.blade.php [hunk 3]
```diff
@@ -40,6 +52,23 @@
 @endif
 
 <p class='smalltext !leading-[14px]'>
-    {{-- Keep this all on a single line so white space isn't added before the comma --}}
-    {{ $formattedPostTimestamp }}@if($formattedEditTimestamp)<span class='italic smalltext !leading-[14px]'>, <span class='hidden sm:inline'>last</span> edited {{ $formattedEditTimestamp }}</span> @endif
+    <span 
+        title="{{ $shouldUsePostedTimeAgoDate ? $postCreatedTimestamp : null }}" 
+        class="{{ $shouldUsePostedTimeAgoDate ? "cursor-help" : null }}"
```
Comment (Jamiras) [inline resources/views/community/components/forum/post-comment-meta.blade.php:57]: This logic ends up with blank title and class attributes. 
```
<span
    title=""
    class=""
>
    
  July 12 2023, 5:22pm    </span>
```

Shouldn't they just be excluded entirely? Will specifying an empty class eliminate any default classes defined in the css?

File: resources/views/community/components/forum/post-comment-meta.blade.php (hunk 2 above)
Comment (Jamiras) [reply resources/views/community/components/forum/post-comment-meta.blade.php:56]: Same comment for forum lists.

Comment (wescopeland) [reply resources/views/community/components/forum/post-comment-meta.blade.php]: Good call, resolved in https://github.com/RetroAchievements/RAWeb/pull/1652/commits/ae22870247e6e877641fa493b7570f0a74d1d5d8.

File: resources/views/community/components/forum/post-comment-meta.blade.php (hunk 3 above)
Comment (wescopeland) [reply resources/views/community/components/forum/post-comment-meta.blade.php]: This was my React.js brain trying to do something with Laravel Blade. In React, passing `null` to an attribute causes that attribute to not be sent to the DOM. But this doesn't seem to be the case for Blade!

After studying for a little bit, I've learned the following syntax in Blade is valid and achieves what we're looking for:

```
<span
    @if($shouldUseEditedTimeAgoDate)
        class="cursor-help"
        title="{{ getNiceDate(strtotime($postEditedTimestamp)) }}"
    @endif
>
```

File: resources/views/community/components/forum/post-comment-meta.blade.php (hunk 2 above)
Comment (Jamiras) [reply resources/views/community/components/forum/post-comment-meta.blade.php]: It's still inconsistent between the tooltip and what's displayed in the text:
image

"July 12" vs "15 Jul" and "5:22pm" vs "15:37"

Comment (wescopeland) [reply resources/views/community/components/forum/post-comment-meta.blade.php]: Ah sorry, I went a little too fast and didn't notice `getNiceDate()` outputted a different format. Should be cleared up in https://github.com/RetroAchievements/RAWeb/pull/1652/commits/86d2636d70f23ee2df3f77528d58a939aeb94e2f.

**Review by Jamiras — APPROVED** [review]

**Comment by wescopeland** [comment]
I've resolved the conflicts from yesterday's merges. The conflict resolution was nontrivial, so it's probably worth another glance by a reviewer before merging.

File: app/Helpers/render/forum.php [hunk 4]
```diff
@@ -43,7 +43,7 @@ function RenderRecentForumPostsComponent(int $numToFetch = 4): void
             echo "<div class='embedded mb-1'>";
             echo "<div class='flex justify-between items-center'><div>";
             echo userAvatar($author, iconSize: 16);
-            echo " <span class='smalldate'>$postedAt</span></div>";
+            echo " <span class='smalldate cursor-help' title='" . Carbon::parse($nextData['PostedAt'])->format('F j Y, g:ia') . "'>$postedAt</span></div>";
```
Comment (Jamiras) [inline app/Helpers/render/forum.php:46]: Tooltips are showing up when 'absolute dates' are enabled.
image

Also in forum post history.

Comment (wescopeland) [reply app/Helpers/render/forum.php]: Thanks, I figured I missed something. Should be resolved in https://github.com/RetroAchievements/RAWeb/pull/1652/commits/fb45f5af58d70395342162b89e197ec157b1f571.

**Review by Jamiras — APPROVED** [review]

**Pull request merged**